include/seastar/core/abort_source.hh

```diff
@@ -115,7 +115,11 @@ public:
     /// Must be called only once, otherwise terminate will be called.
     void request_abort() noexcept {
         auto subs = std::exchange(_subscriptions, std::nullopt).value();
-        subs.clear_and_dispose([] (subscription* s) { s->on_abort(); });
+        while (!subs.empty()) {
+            subscription& s = subs.front();
+            s.unlink();
+            s.on_abort();
+        }
```

Comment:
Isn't this exactly what `erase_and_dispose()` does (not clear_and_dispose, but rather erase_and_dispose)?

(I'm sorry if I'm asking a stupid question, but as usual, the Boost documentation on this - https://www.boost.org/doc/libs/1_43_0/doc/html/intrusive/erasing_and_disposing.html - is absolutely useless).

Reply:
No, not exactly. Apparently `boost::intrusive::erase_and_dispose` suffers from the same issue as `clear_and_dispose` - if disposing an element in the range will (auto-)unlink the next element in the list.

Reply:
I still don't understand what you mean "unlink the next element in the list". What can erase_and_dispose do differently from what you do, namely unlinking each element right before calling the dispose function? Did you try it?
Here is what I see in the Boost source code:
```
   iterator erase_and_dispose(const_iterator i, Disposer disposer)
   {
      node_ptr to_erase(i.pointed_node());
      ++i;
      node_algorithms::unlink(to_erase);
      this->priv_size_traits().decrement();
      if(safemode_or_autounlink)
         node_algorithms::init(to_erase);
      disposer(this->priv_value_traits().to_value_ptr(to_erase));
      return i.unconst();
```
I don't understand the details, but it also seems to first unlink and then call the disposer function. I don't know what the init() function does or whether we call it in this case - is this what bothers you?